src/lib/prov/tpm2/tpm2_crypto_backend/tpm2_crypto_backend.cpp

```diff
@@ -0,0 +1,847 @@
+
+      return TSS2_RC_SUCCESS;
+   });
+   #else
+   BOTAN_UNUSED(pub_tpm_key, in_size, in_buffer, max_out_size, out_buffer, label, userdata);
+   return TSS2_ESYS_RC_NOT_IMPLEMENTED;
+   #endif
+}
+
+/** Computation of an ephemeral ECC key and shared secret Z.
+ *
+ * According to the description in TPM spec part 1 C 6.1 a shared secret
+ * between application and TPM is computed (ECDH). An ephemeral ECC key and a
+ * TPM key are used for the ECDH key exchange.
+ * @param[in] key The key to be used for ECDH key exchange.
+ * @param[in] max_out_size the max size for the output of the public key of the
+ *            computed ephemeral key.
+ * @param[out] Z The computed shared secret.
+ * @param[out] Q The public part of the ephemeral key in TPM format.
+ * @param[out] out_buffer The public part of the ephemeral key will be marshaled
+ *             to this buffer.
+ * @param[out] out_size The size of the marshaled output.
+ * @param[in,out] userdata information.
+ * @retval TSS2_RC_SUCCESS on success
+ * @retval USER_DEFINED user defined errors on failure.
+ */
+static TSS2_RC get_ecdh_point(TPM2B_PUBLIC* key,
+                              size_t max_out_size,
+                              TPM2B_ECC_PARAMETER* Z,
+                              TPMS_ECC_POINT* Q,
+                              BYTE* out_buffer,
+                              size_t* out_size,
+                              void* userdata) {
+   *out_size = 0;
+   #if defined(BOTAN_HAS_ECDH)
+   return thunk([&] {
+      BOTAN_ASSERT_NONNULL(key);
+
+      auto ccs = get(userdata);
+      if(!ccs) {
+         return TSS2_ESYS_RC_BAD_REFERENCE;
+      }
+
+      Botan::RandomNumberGenerator& rng = *ccs->get().rng;
+
+      // 1: Get TPM public key
+      const auto tpm_pub_inputs = Botan::TPM2::ecc_pubkey_from_tss2_public(key);
+      const auto tpm_sw_pubkey = Botan::ECDH_PublicKey(tpm_pub_inputs.first, tpm_pub_inputs.second);
+
+      const auto curve_order_byte_size = tpm_sw_pubkey.domain().get_p_bytes();
+
+      // 2: Generate ephemeral key
+      const auto eph_key = Botan::ECDH_PrivateKey(rng, tpm_sw_pubkey.domain());
+
+      // Get public key coordinates
+      const Botan::PointGFp& eph_pub_point = eph_key.public_point();
+      const auto& bn_x = eph_pub_point.get_affine_x();
+      const auto& bn_y = eph_pub_point.get_affine_y();
+
+      // Serialize affine coordinates into TPM2B_ECC_PARAMETER with Big Endian encoding.
+      // This ensures bn_{x,y}.bytes() <= curve_order_byte_size.
+      Botan::TPM2::copy_into<TPM2B_ECC_PARAMETER>(Q->x, bn_x.serialize(curve_order_byte_size));
+      Botan::TPM2::copy_into<TPM2B_ECC_PARAMETER>(Q->y, bn_y.serialize(curve_order_byte_size));
+
+      // 3: ECDH Key Agreement
+      Botan::PK_Key_Agreement ecdh(eph_key, rng, "Raw" /*No KDF used here*/);
+      const auto shared_secret = ecdh.derive_key(0 /*Ignored for raw KDF*/, tpm_sw_pubkey.public_value()).bits_of();
```

Comment:
[Continuing the `generate_another()` discussion](https://github.com/Rohde-Schwarz/botan/pull/14#discussion_r1785082889): There is a way to use it here, like so:

```suggestion
      // 2: Generate ephemeral key
      const auto eph_key = tpm_sw_pubkey.generate_another(rng);

      // Get public key coordinates
      const auto eph_pub_value = eph_key->raw_public_key_bits();
      Botan::BufferSlicer slicer(eph_pub_value);
      BOTAN_ASSERT_NOMSG(slicer.take_byte() == 0x04);  // Uncompressed point

      // Serialize affine coordinates into TPM2B_ECC_PARAMETER with Big Endian encoding.
      // This ensures bn_{x,y}.bytes() <= curve_order_byte_size.
      Botan::TPM2::copy_into(Q->x, slicer.take(curve_order_byte_size));
      Botan::TPM2::copy_into(Q->y, slicer.take(curve_order_byte_size));
      BOTAN_ASSERT_NOMSG(slicer.empty());

      // 3: ECDH Key Agreement
      Botan::PK_Key_Agreement ecdh(*eph_key, rng, "Raw" /*No KDF used here*/);
      const auto shared_secret = ecdh.derive_key(0 /*Ignored for raw KDF*/, tpm_sw_pubkey.public_value()).bits_of();
```

... for key agreement keys, `raw_public_key_bits()` returns the "public value" representation. And for ECC this is the "uncompressed point format" by default (i.e. `0x04 || pubpoint.x || pubpoint.y`). So you don't really need a cast here, but you'd have to depend on the "default point encoding" and split this up into `Q->x` and `Q->y`.

Note: I think explicitly calling the `ECDH_PrivateKey`, as you did, is the better approach. No need to abstract the type of private key here, as it will always be ECDH. This is just for the sake of argument.